### ISSUE TITLE
initialSyncLimit of MatrixClient.startClient() doesn't behave literally.

### ISSUE THREAD

**Issue by leonlianght** [matrix-org/matrix-js-sdk#715]
**TL;DR**

Literally,`initialSyncLimit` option means that the limit events of the initial sync request, it would create a filter for the first sync, so the total events number of one room for the first sync would be this option.
However, after the first sync, it would keep using the same filter for every sync request.
https://github.com/matrix-org/matrix-js-sdk/blob/4d310cd4618db4e98a8e6b5eb812480102ee4dee/src/sync.js#L462-L470
The above code will generate a timeline limit filter(if the filter option is not set) and this filter would be used in all the sync requests.
I wonder if it is intended or a miss for this kind of situation.

------

I have met a situation that when an account contains large amount of rooms (1000+ in my case), the size of first sync request(without filter) would be around 10MB, which would take more than 60s to finish the request. It's way too long for a user to wait a few minutes to finish initial login, so I tried to find a way to improve it.

The first thing came to my mind is 'lazy loading'. if I am able to get the room list first(first sync), and load the timelines after the room list is ready, the size of the first sync could be significantly reduced.

When I tried to find a way to improve it, I found the `initialSyncLimit ` option of `MatrixClient.startClient()`:
https://github.com/matrix-org/matrix-js-sdk/blob/4d310cd4618db4e98a8e6b5eb812480102ee4dee/src/client.js#L3014-L3015
What I did is to set this `initialSyncLimit` to 0, as I expected, the sync won't give me timeline and the size of this sync reduce to < 2MB and the it only takes 10s, it is much better than waiting for a minute!

And then I code a 'lazy loading' function to retrieve the timeline for a room, although it would took a looooooooong time to fetch all the timelines, the user can send and receive messages during this time.

Everything looks fine until I found an issue when a new message is received, the live timeline would be reset. After digging around I found the `initialSyncLimit` filter is applied for every sync requests, that's quite wired since I thought this option would only applied for the first sync.

----

Another thing I don't quite understand why is a `limit sync` needed, going through the docs/code/comments but still not fully understand :(
https://github.com/matrix-org/matrix-js-sdk/blob/4d310cd4618db4e98a8e6b5eb812480102ee4dee/src/sync.js#L1003-L1013
